RecoParticleFlow/PFClusterProducer/plugins/alpaka/PFClusterSoAProducerKernel.dev.cc

```diff
@@ -0,0 +1,1634 @@
+      }
+    }
+    alpaka::syncBlockThreads(acc);  // all threads call sync
+
+    int rhThreadIdx = -1;
+    Position4 rhThreadPos;
+    if (tid < (nRHNotSeed - 1)) {
+      rhThreadIdx = rechits[tid];  // Index when thread represents rechit
+      rhThreadPos = Position4{pfRecHits[rhThreadIdx].x(), pfRecHits[rhThreadIdx].y(), pfRecHits[rhThreadIdx].z(), 1.};
+    }
+
+    // Neighbors when threadIdx represents seed
+    int seedThreadIdx = -1;
+    Neighbours4 seedNeighbors = Neighbours4{-9, -9, -9, -9};
+    float seedEnergy = -1.;
+    Position4 seedInitClusterPos = Position4{0., 0., 0., 0.};
+    if (tid < nSeeds) {
+      if constexpr (debug)
+        printf("tid: %d\n", tid);
+      seedThreadIdx = getSeedRhIdx(seeds, tid);
+      seedNeighbors = Neighbours4{pfRecHits[seedThreadIdx].neighbours()(0),
+                                  pfRecHits[seedThreadIdx].neighbours()(1),
+                                  pfRecHits[seedThreadIdx].neighbours()(2),
+                                  pfRecHits[seedThreadIdx].neighbours()(3)};
+      seedEnergy = pfRecHits[seedThreadIdx].energy();
+
+      // Compute initial cluster position shift for seed
+      computeClusterPos(pfClusParams, seedInitClusterPos, 1., seedThreadIdx, pfRecHits, rhENormInv);
+    }
+
+    do {
+      if constexpr (debug) {
+        if (once_per_block(acc))
+          printf("\n--- Now on iter %d for topoId %d ---\n", iter, topoId);
+      }
+
+      // Reset rhFracSum
+      rhFracSum[tid] = 0.;
+      if (once_per_block(acc))
+        diff2 = -1;
+
+      if (tid < (nRHNotSeed - 1)) {
+        for (int s = 0; s < nSeeds; s++) {
+          float dist2 = (clusterPos[s].x - rhThreadPos.x) * (clusterPos[s].x - rhThreadPos.x) +
+                        (clusterPos[s].y - rhThreadPos.y) * (clusterPos[s].y - rhThreadPos.y) +
+                        (clusterPos[s].z - rhThreadPos.z) * (clusterPos[s].z - rhThreadPos.z);
+
+          float d2 = dist2 / pfClusParams.showerSigma2();
+          float fraction = clusterEnergy[s] * rhENormInv * expf(-0.5 * d2);
+
+          rhFracSum[tid] += fraction;
+        }
+      }
+      alpaka::syncBlockThreads(acc);  // all threads call sync
+
+      if (tid < (nRHNotSeed - 1)) {
+        for (int s = 0; s < nSeeds; s++) {
+          int i = seeds[s];
+          float dist2 = (clusterPos[s].x - rhThreadPos.x) * (clusterPos[s].x - rhThreadPos.x) +
+                        (clusterPos[s].y - rhThreadPos.y) * (clusterPos[s].y - rhThreadPos.y) +
+                        (clusterPos[s].z - rhThreadPos.z) * (clusterPos[s].z - rhThreadPos.z);
+
+          float d2 = dist2 / pfClusParams.showerSigma2();
+          float fraction = clusterEnergy[s] * rhENormInv * expf(-0.5 * d2);
+
+          if (rhFracSum[tid] > pfClusParams.minFracTot()) {
+            float fracpct = fraction / rhFracSum[tid];
+            if (fracpct > 0.9999 || (d2 < 85. && fracpct > pfClusParams.minFracToKeep())) {
```

Comment:
This `85.` should be changed to `100.` as the other functions, I think that specific value is a remnant of development. As for what this value represents, I believe it is a form of distance cutoff for energy sharing. I think that `showerSigma2` can be used in place of `100.` and this would make the condition for energy sharing that the distance from seed to rechit is less than the Gaussian width squared. Does this sound right @hatakeyamak?

For the `0.9999`, this predates me but looks to me like a check that any rechit fraction that is essentially `1` is kept.

Reply:
Looking at `RecoParticleFlow/PFClusterProducer/plugins/Basic2DGenericPFlowClusterizer.cc`, I see https://github.com/cms-sw/cmssw/blob/754d4fbc8bf3e05df6de16937fe6f4d30f5d8699/RecoParticleFlow/PFClusterProducer/plugins/Basic2DGenericPFlowClusterizer.cc#L259-L262 and https://github.com/cms-sw/cmssw/blob/754d4fbc8bf3e05df6de16937fe6f4d30f5d8699/RecoParticleFlow/PFClusterProducer/plugins/Basic2DGenericPFlowClusterizer.cc#L293-L294 So perhaps I will just keep `100.` and add a comment that it is the distance cutoff. And also add a comment that `0.9999` is a quick check to keep all rechits which contribute all of their energy to the cluster.